### ISSUE TITLE
AemEnvironmentIndicatorFilter fails to trigger on the author environment, due to WCMMode being DISABLED

### ISSUE THREAD

**Issue by henrykuijpers** [Adobe-Consulting-Services/acs-aem-commons#2402]
## Bug Report

**Current Behavior**
When configuring the AemEnvironmentIndicatorFilter with the configuration: `css-color="#ff00ff"`, we do not see an environment indicator being shown in the OSGi console, in the AEM interface, etc. 

**Expected behavior/code**
An environment indicator being shown.

**Environment**
- AEM 6.4 SP8
- ACS AEM Commons 4.8.0
- JRE version 8 build 200+

**Possible Solution**
It looks like the WCMRequestFilter is being ran after the AemEnvironmentIndicatorFilter. Therefore the AemEnvironmentIndicatorFilter receives WCMMode.DISABLED as the current WCMMode, which is the default excluded WCMMode.

**Comment by henrykuijpers** [comment]
It looks like the filter is being registered as a Jetty filter, while the WCMRequestFilter is a Sling Filter and therefore runs much later in the chain.

It also looks like a recent update changed the way of obtaining the WCMMode from request.getAttribute(...) to WCMMode.fromRequest(request).

https://github.com/Adobe-Consulting-Services/acs-aem-commons/blob/master/bundle/src/main/java/com/adobe/acs/commons/wcm/impl/AemEnvironmentIndicatorFilter.java#L371

The problem here is that in the previous case, null was returned, but now WCMMode.DISABLED is returned (since WCMMode.fromRequest()) actually returns WCMMode.DISABLED when the attribute could not be found. Therefore, it never registers the filter.

WDYT @davidjgonzalez @justinedelson ?

**Comment by rubnig** [comment]
If you change it to a Sling Filter instead of a Jetty Filter, you will also get the following problem: the Indicator won't show up in CRX/DE Publish (which it did before) because the WCMMode == WCMMode.DISABLED in that case too.

It looks like the Unit Test is also testing the class not as it should be, because it is using a spy: https://github.com/Adobe-Consulting-Services/acs-aem-commons/blob/master/bundle/src/test/java/com/adobe/acs/commons/wcm/impl/AemEnvironmentIndicatorFilterTest.java#L76

Maybe the test can also be updated with actual testing of the accepts method.

**Comment by joerghoh** [comment]
Hi @rubnig, thanks for reporting. I will try to investigate on that.
As you already mentioned, this filter has to be registered with the HTTP whiteboard, otherwise it won't appear on the Webconsole and in CRXDE.

**Comment by joerghoh** [comment]
@rubnig I tried to reproduce your problem, but the information provided is not sufficient. Can you please be more specific which OSGI property I have to set with the value  ```css-color="#ff00ff"```?

If I follow the documentation at https://adobe-consulting-services.github.io/acs-aem-commons/features/environment-indicator/index.html, it works fine. 

Can you please refine your problem description, so I can actually reproduce it?

**Comment by henrykuijpers** [comment]
Hi @joerghoh , did you use the 4.8.0 version in combination with for example AEM 6.4.8? 

If you configure it, it might work. If you however restart AEM, it's not working anymore. 

While debugging, you can see that the `WCMMode.fromRequest(request)` yields `WCMMode.DISABLED`.

The problem is that previously, the method that was determining the WCMMode, was returning `null` in that case.

**Comment by joerghoh** [comment]
hi @henrykuijpers 

I tried this approach:
* I had 6.4.5 handy.
* I deployed ACS AEM Commons 4.8.5-SNAPSHOT
* configured the EnvironmentIndicator like this
  * set the property "ccs-override" to the CSS posted on the documentation page
  * set the property "inner-html"
* refreshed the page, the environment indicator is present in the OSGI webconsole
* I restarted the instance to deploy 6.4.8, afterwards the environment indicator is present.
* restarted the instance again, the environment indicator is present.

From what I see the last change to the EnvironmentIndicatorFilter has been with #2271 fixing a bug which seems quite unrelated to your problem.

From looking at the code, the WCMMode.DISABLED (if no WCMMode is specified) is not problematic per se.

Do you have set the OSGI property "excluded-wcm-modes" to contain "DISABLED"? In that case you don't get the EnvironmentIndicator set, that's true. But that's going to be expected then.

**Comment by henrykuijpers** [comment]
We'll need to test this a bit further tomorrow then. Can you in the meantime please try with 6.4.8? As that is the exact version we were using when we saw it not working anymore.

**Comment by joerghoh** [comment]
I have version 6.4.8 running now, and everything is working fine.

**Comment by henrykuijpers** [comment]
We'll investigate further tomorrow then. 

How can WCMMode be available as a request attribute in a jetty filter, if it is being set in the WCMRequestFilter, which is a sling filter and therefore runs much later? 
Because, if it is not available a request attribute, `WCMMode.fromRequest(request)` returns `WCMMode.DISABLED`. At least, that's what we saw from decompiled code. And since the default excluded WCMMode is DISABLED, it does not render the indicator. 

Do you see a value other than `DISABLED` come back when `getWcmMode` is being called? @joerghoh

**Comment by joerghoh** [comment]
Good catch. 

I agree, the WCMMode is never available as request attribute because the environment filter is running earlier than the WCMRequestFilter. Thus any handling based on the wcmMode is unlikely to function (not sure if it ever worked at all), except the case when you provide "DISABLED" to the "excluded-wcm-modes" property.

**Comment by henrykuijpers** [comment]
Hi @joerghoh ,

So I'm using ACS AEM Commons 4.8.0, with the following configuration:
```
<?xml version="1.0" encoding="UTF-8"?>
<jcr:root xmlns:sling="http://sling.apache.org/jcr/sling/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0"
    jcr:primaryType="sling:OsgiConfig"
    css-color="#ff7676"/>
```

This is my debugger view when I'm in the `accepts`-method:
![image](https://user-images.githubusercontent.com/6546957/92378628-5e280400-f106-11ea-8974-5a655d9781df.png)
As you can see, I have the `css`-field set (css-color OSGi property) and I'm using the default excluded WCM modes.

My debugger view when I'm in WCMMode.fromRequest:
![image](https://user-images.githubusercontent.com/6546957/92378737-99c2ce00-f106-11ea-9a2c-310c8083da59.png)
![image](https://user-images.githubusercontent.com/6546957/92378774-a6dfbd00-f106-11ea-97eb-72834573b787.png)
As you can see, when the request attribute could not be found (which makes sense, since that request attribute is set in `WCMRequestFilter`,  which runs *after* the `AemEnvironmentIndicatorFilter`, so it can never be set at this point) it defaults to `WCMMode.DISABLED`.

Then, of course, this method will return `true`:
```
    boolean isDisallowedWcmMode(WCMMode currentMode, String[] excludedWcmModes) {
        return currentMode == null || StringUtils.equalsAnyIgnoreCase(currentMode.name(), excludedWcmModes);
    }
```
and therefore the environment indicator will never be shown. WDYT @joerghoh ?

**Comment by joerghoh** [comment]
Hi @henrykuijpers 

I am bit confused, because in see in your screenshot, that `titlePrefix` is empty (""), and thus `isImproperlyConfigured`() would return false already, thus returning immediately.

And regarding the method "`sDisallowedWcmMode()`: I agree, that this logic is unlikely to work properly. It just works, if **no** WcmMode is configured to be excluded.

Can you give it a try with that? 

(Yes, we need to fix this logic, but I doubt that we can do that properly without duplicating the wcm-mode logic from AEM. I would rather remove "do not display it in certain runmodes" feature completely.)

**Comment by henrykuijpers** [comment]
@joerghoh , with regards to `isImproperlyConfigured`, no that is not true: if you look carefully you'll see that "improperly configured" means that both the CSS color and the title prefix are empty. In our case we do not set the title prefix, as we do not want to modify the browser title. 

We can indeed try with the disallowed WCMModes being an empty array, but that will also render the indicator on `WCMMode.DISABLED`, which is obviously not what we want.

**Comment by joerghoh** [comment]
OK, got it. 

The only way of fix I can think of would be to use an additional filter, which is run after the WcmMode is determined, and which remove the data inserted by the first filter (which is still registered at the HTTP Bridge as today).

**Comment by kwin** [comment]
This approach would require buffering the whole response. I would rather get rid of the option to enable/disable based on WCM mode (i.e. remove property https://github.com/Adobe-Consulting-Services/acs-aem-commons/blob/eb7219a8ac462414b51e3e76cf74bbdbefbe78bb/bundle/src/main/java/com/adobe/acs/commons/wcm/impl/AemEnvironmentIndicatorFilter.java#L149) as this was also never documented at https://adobe-consulting-services.github.io/acs-aem-commons/features/environment-indicator/index.html. Instead I would introduce options to disable for certain path prefixes (/content) and enable for others (i.e. /editor, /sites). That would also cover the "Show as published" option which would not expose the indicator any longer.

**Comment by joerghoh** [comment]
@kwin The buffering is already there :-(, see https://github.com/Adobe-Consulting-Services/acs-aem-commons/blob/eb7219a8ac462414b51e3e76cf74bbdbefbe78bb/bundle/src/main/java/com/adobe/acs/commons/wcm/impl/AemEnvironmentIndicatorFilter.java#L183

and I don't really like that either, and I also don't see a way to get rid of this behavior without changing the complete approach.

But it's actually not that hard to implement the proper handling of the WcmModes with the mentioned 2 filters.

**Comment by joerghoh** [comment]
@henrykuijpers Can you please try https://github.com/Adobe-Consulting-Services/acs-aem-commons/pull/2407 and check if it fixes the described problem?
In my tests it does, but it's actually hard to write unittests for this, because it relies a lot on behavior of AEM in other areas.

**Comment by henrykuijpers** [comment]
This is a pretty good solution, @joerghoh . I'll make some time tomorrow to test this. :) 

I've added some comments in the PR, out of which one is pretty important. 

I'd like to highlight something though: We've never configured the excluded wcm modes ourselves. This was default behavior. Previously, this was working (although it was a big fat bug). Recently, the way of obtaining the WCMMode was changed in the filter code, which made it return DISABLED instead of null. That's when the indicator stopped working for us (and probably almost everyone else). We were not using some sort of exotic configuration. It's the recent code change that broke the filter. 

I still don't like the way this filter is unit-tested, the spy-construction that is used prevents important runtime functionality to be properly tested (the WCMMode.disabled issue would have been found a lot earlier with proper rest cases in place), but let's fix one thing at a time. I'd be willing to improve the unit tests together with my colleague in another issue.

**Comment by henrykuijpers** [comment]
I thought it was strange that you didn't remove the WCMMode-check from the `accepts`-method. Today I had a chance to test the changes and indeed, with that check still in place, `accepts` would return false when accessing crx/de for example. 

So, when I deployed the changed code and then accessed the OSGi console, crx/de, etc. I didn't see an environment indicator. When stepping through the code, it became apparent immediately that the WCMMode check was still in place outside the Sling Filter (where it could work).

Is my environment somehow different? The lack of the environment indicator was spotted immediately in my case.

I suggested another change in the PR.
